[CDAP-20872] Add AeadCipher SPI for Encryption in CDAP

dli357 opened the pull request:
This PR adds a new SPI for supporting encryption using Tink. The following encryption implementations are supported:
  - Tink Cleartext: Uses a file-based Tink key read from SConfiguration.
  - Tink GCP Envelope: Uses [Tink client-side envelope encryption](https://developers.google.com/tink/client-side-encryption) with GCP Cloud KMS as the key management service.

Additionally, the PR also uses the new encryption SPI in the following ways:
  - Adds data storage encryption to CredentialIdentityStore and CredentialProfileStore
  - Refactors the existing user credential encryption in AuthenticationHandler and NettyRouter to use the new SPI

Finally, this PR also removes the Tink dependency from cdap-security, along with protobuf and tink shading.

chtyim commented on cdap-security/src/main/java/io/cdap/cdap/security/encryption/guice/DefaultAeadCipherService.java:
```diff
@@ -0,0 +1,65 @@
+
+  @Override
+  protected void startUp() throws Exception {
+    aeadCipher.initialize(new AeadCipherContext(properties, secureProperties));
```
Does it need to be a Service or can it be just a one time on demand initialization? Using Service making the usage a little bit more involving.

dli357 replied:
Changed to be lazily initialized instead of initialized as a service

dli357 committed:
[CDAP-20872] Create AeadCipher SPI for supporting extensible encryption in CDAP

[CDAP-20872] Add data encryption for CredentialIdentityStore and CredentialProfileStore

[CDAP-20872] Refactor user credential encryption to use new SPI

[CDAP-20872] Remove Tink dependency from cdap-security

[CDAP-20872] Addressed comments, renamed AeadCipher SPI to AeadCipherCryptor

[CDAP-20872] Change CipherException to extend RuntimeException and address comments

[CDAP-20872] Fix checktstyle

albertshau commented on cdap-encryption-ext-tink/src/main/java/io/cdap/cdap/encryption/tink/CloudKmsClient.java:
```diff
@@ -0,0 +1,180 @@
+   * Get project id from the metadata server. Makes a request to the metadata server that lives on
+   * the VM, as described at https://cloud.google.com/compute/docs/storing-retrieving-metadata.
+   */
+  private String getSystemProjectId(String metadataServerApi) throws IOException {
```
any reason we need to do this manually instead of ServiceOptions.getDefaultProjectId()?

That way would also work when CDAP is running in non-GCP environments (like sandbox)

dli357 replied:
This is remnants of another class I copied this class from in the Secure Store extension. Good call though, I went ahead and added the logic to read the project ID using `ServiceOptions.getDefaultProjectId()` if it's not set in CConf.

albertshau commented on cdap-encryption-ext-tink/src/main/java/io/cdap/cdap/encryption/tink/CloudKmsClient.java:
```diff
@@ -0,0 +1,180 @@
+  /**
+   * Creates a new key ring with the given id.
+   *
+   * @throws IOException if there's an error while creating the key ring
```
fix javadoc

dli357 replied:
Done

albertshau commented on cdap-encryption-ext-tink/src/main/java/io/cdap/cdap/encryption/tink/CloudKmsClient.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    LocationName locationName = LocationName.of(projectId, location);
+    client.createKeyRing(locationName, keyRingId, KeyRing.newBuilder().build());
```
should catch ApiException here too. At least to check if it's a 409 which can be ignored, probably want to translate it to a CipherInitializationException for other errors

dli357 replied:
Done

albertshau commented on cdap-encryption-ext-tink/src/main/java/io/cdap/cdap/encryption/tink/CloudKmsClient.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    // Calculate the date 24 hours from now (this is used below).
+    long tomorrow = java.time.Instant.now().plus(24, ChronoUnit.HOURS).getEpochSecond();
```
import java.time.Instant

dli357 replied:
Done

albertshau commented on cdap-security/src/main/java/io/cdap/cdap/security/encryption/NoOpAeadCipher.java:
```diff
@@ -0,0 +1,34 @@
+ * Performs no encryption.
+ */
+public class NoOpAeadCipher implements
+    AeadCipher {
```
nit: don't need line wrap

dli357 replied:
Done

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java:
```diff
@@ -93,15 +127,21 @@ public Collection<CredentialIdentityId> listForProfile(StructuredTableContext co
             id.getName()));
-    return table.read(key).map(row -> GSON.fromJson(row
-        .getString(CredentialProviderStore.IDENTITY_DATA_FIELD), CredentialIdentity.class));
+    Optional<StructuredRow> row = table.read(key);
```
Probably worth to have a private method to do the raw read such that you don't need to repeat the same logic here and in the exists method.

dli357 replied:
Done

albertshau commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialProfileStore.java:
```diff
@@ -38,14 +42,26 @@
+  private final AeadCipher dataStorageCipher;
+
+  @Inject
+  public CredentialProfileStore(@Named(DataStorageAeadEncryptionModule.DATA_STORAGE_ENCRYPTION)
```
injected constructors should generally be package private to prevent manual construction

dli357 replied:
It was originally needed since the stores were in their own directory, but I moved it to the same directory as test classes since that seems unnecessary. Done

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java:
```diff
@@ -38,14 +42,26 @@
  */
 public class CredentialIdentityStore {
 
+  private static final byte[] CREDENTIAL_IDENTITY_STORE_AD = CredentialIdentityStore.class
```
If we refactor and changed the class name later, will it break when trying to decrypt data that was encrypted using the old class name as the ad?

chtyim replied:
I think it is safer to actually store the ad being used for encryption in the structured table as well so that we can use it for decryption.

dli357 replied:
Storing the AD in the table removes the protection against an attacker who has write access to the table; in particular, they could lift a value encrypted with a different AD from another table and write it to the current table, and it will decrypt properly.

Your original concern is valid though; I think it's probably better to hardcode this as a string value with a note that changing the value is backwards incompatible.

chtyim replied:
Because we are using the same key to encrypt/decrypt data in different tables?

dli357 replied:
Yup

chtyim commented on cdap-security/src/main/java/io/cdap/cdap/security/encryption/AeadCipher.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * An interface intended to represent a service providing AEAD symmetric encryption primitives.
+ */
+public interface AeadCipher {
```
It's not a good idea to have two different interfaces with the same name in two different packages. It's a bit confusing to read the code.

dli357 replied:
Done, renamed SPI AeadCipher to AeadCipherCryptor.

chtyim commented on cdap-security/src/main/java/io/cdap/cdap/security/encryption/AeadCipher.java:
```diff
@@ -0,0 +1,104 @@
+   * @return encrypted data in base64 encoded form
+   * @throws CipherOperationException if encryption fails
+   */
+  default String encryptStringToBase64(String plainData, byte[] associatedData)
```
You can just call it `encryptToBase64` as the parameter type already tell what the plaintext type is.

dli357 replied:
Done

chtyim commented on cdap-security/src/main/java/io/cdap/cdap/security/encryption/guice/LazyDelegateAeadCipher.java:
```diff
@@ -0,0 +1,68 @@
+
+  @Override
+  public byte[] encrypt(byte[] plainData, byte[] associatedData) throws CipherException {
+    if (!initialized) {
```
This would not be concurrent safe. It's better to make it safe by using a proper double sync singleton technique.

dli357 replied:
Good catch, done

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityStore.java:
```diff
@@ -93,15 +125,15 @@ public Collection<CredentialIdentityId> listForProfile(StructuredTableContext co
-        .getString(CredentialProviderStore.IDENTITY_DATA_FIELD), CredentialIdentity.class));
+      throws CipherException, IOException {
+    Optional<StructuredRow> row = readIdentity(context, id);
+    if (!row.isPresent()) {
```
Calling isPresent followed by transformation is not considered a good practice. In this case, it's better be:

```java
return readIdentity(context, id)
  .map(r -> r.getBytes(CredentialProviderStore.IDENTITY_DATA_FIELD))
  .map(b -> new String(b, UTF_8))
  .map(s -> GSON.fromJson(s, CredentialIdentity.class));
```

dli357 replied:
Done

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialProfileStore.java:
```diff
@@ -74,15 +111,21 @@ public Collection<CredentialProfileId> list(StructuredTableContext context, Stri
-    return table.read(key).map(row -> GSON.fromJson(row
-        .getString(CredentialProviderStore.PROFILE_DATA_FIELD), CredentialProfile.class));
+    Optional<StructuredRow> row = table.read(key);
+    if (!row.isPresent()) {
```
Similar to the comment above, it's better to use map instead of isPresent

dli357 replied:
Done

chtyim commented on cdap-encryption-ext-tink/pom.xml:
```diff
@@ -0,0 +1,164 @@
+      <dependency>
+        <groupId>com.google.cloud</groupId>
+        <artifactId>libraries-bom</artifactId>
+        <version>26.23.0</version>
```
nit: move this version to the properties too.

dli357 replied:
Done

chtyim commented on cdap-security/src/main/java/io/cdap/cdap/security/encryption/guice/LazyDelegateAeadCipher.java:
```diff
@@ -0,0 +1,70 @@
+
+  private void lazyInitialize() throws CipherInitializationException {
+    if (!initialized) {
+      synchronized(this) {
```
Please fix the checkstyle

dli357 replied:
Done

chtyim commented on cdap-security/src/main/java/io/cdap/cdap/security/encryption/guice/LazyDelegateAeadCipher.java:
```diff
@@ -0,0 +1,70 @@
+    this.aeadCipherCryptor = aeadCipherCryptor;
+    this.properties = properties;
+    this.secureProperties = secureProperties;
+    this.initialized = false;
```
nit: Not needed.

dli357 replied:
Done

chtyim commented on cdap-security/src/main/java/io/cdap/cdap/security/encryption/guice/LazyDelegateAeadCipher.java:
```diff
@@ -0,0 +1,70 @@
+      Map<String, String> secureProperties) {
+    this.aeadCipherCryptor = aeadCipherCryptor;
+    this.properties = properties;
+    this.secureProperties = secureProperties;
```
Make an immutable copy of incoming maps.

dli357 replied:
Done

chtyim reviewed (approved):
LGTM, just couple minor comments left.

dli357 commented:
Thank you for the reviews!
